lib/errors.js

```diff
@@ -285,6 +285,9 @@ internals.annotate = function (stripColorCodes) {
                         delete ref[replacement];
                     }
                     else {
+                        if (typeof ref === 'string') {
```

Comment:
Sorry, I should have been more accurate, while this is fixing the failure, it's not fixing the issue. If you look at the error annotation, it's showing the list of errors correctly but can't pinpoint it on the input.

Eg. with your 2nd example :
```
ValidationError: {
  "a": "{\"b\":{\"c\":{\"d\":{}}}}" // There should be a [1] here or on the parsed object
}

[1] "d" must be a string
```